marconi-sidechain/test/Spec/Marconi/Sidechain/Utils.hs

```diff
@@ -0,0 +1,155 @@
+import Data.Function ((&))
+import Data.List (isInfixOf)
+import Data.List qualified as L
+import Data.Monoid (getLast)
+import Data.Text qualified as T
+import GHC.IO qualified as IO
+import Hedgehog qualified as H
+import Hedgehog.Extras qualified as H
+import Hedgehog.Extras.Internal.Plan qualified as H
+import Hedgehog.Extras.Stock qualified as OS
+import Helpers qualified as Help
+import System.Directory qualified as IO
+import System.Environment qualified as IO
+import System.FilePath qualified as IO
+import System.FilePath.Posix ((</>))
+import System.IO qualified as IO
+import System.Process qualified as IO
+
+-- | Wait up to a specific duration (seconds) for text file to contain all string values
+waitForLog :: FilePath -> Int -> [String] -> IO Bool
+waitForLog logFile t expectedStrings = go (t + 1) (t + 1)
+  where
+    go t1 t2
+      | t1 <= 0 = return False
+      | t1 == t2 = do
+          IO.threadDelay 100_000 -- wait 100ms before initial check
+          go (t1 - 1) t2
+      | otherwise = do
+          contents <- IO.readFile logFile
+          if all (`isInfixOf` contents) expectedStrings
+            then return True
+            else do
+              IO.threadDelay 1_000_000 -- wait 1 second after each subsequent check
+              go (t1 - 1) t1
+
+integrationTest :: H.Integration () -> H.Property
+integrationTest =
+  H.withShrinks 0
+    . H.withTests 1
+    . H.propertyOnce
+    . (liftIO Help.setDarwinTmpdir >>)
+    . H.runFinallies
+
+-- Capture handle contents in a separate thread (e.g. stderr)
+captureHandleContents :: IO.Handle -> IO BSL.ByteString
+captureHandleContents handle = do
+  mvar <- IO.newEmptyMVar
+  _ <- IO.forkIO $ do
+    contents <- BSL.hGetContents handle
+    IO.putMVar mvar contents
+  IO.takeMVar mvar
+
+-- -- | Compute the path to the binary given a package name or an environment variable override.
+binFlex
+  :: String
+  -- ^ Package name
+  -> String
+  -- ^ Environment variable pointing to the binary to run
+  -> IO FilePath
+  -- ^ Path to executable
+binFlex pkg binaryEnv = do
+  maybeEnvBin <- liftIO $ IO.lookupEnv binaryEnv
+  case maybeEnvBin of
+    Just envBin -> return envBin
+    Nothing -> binDist pkg
+
+addExeSuffix :: String -> String
+addExeSuffix s =
+  if ".exe" `L.isSuffixOf` s
+    then s
+    else s <> if OS.isWin32 then ".exe" else ""
+
+-- -- | Find the nearest plan.json going upwards from the current directory.
+findDefaultPlanJsonFile :: IO FilePath
+findDefaultPlanJsonFile = IO.getCurrentDirectory >>= go
+  where
+    go :: FilePath -> IO FilePath
+    go d = do
+      let file = d </> "dist-newstyle/cache/plan.json"
+      exists <- IO.doesFileExist file
+      if exists
+        then return file
+        else do
+          let parent = IO.takeDirectory d
+          if parent == d
+            then return "dist-newstyle/cache/plan.json"
+            else go parent
+
+-- -- | Discover the location of the plan.json file.
+planJsonFile :: String
```

Comment:
Should be IO as return type no?

Reply:
I can remove `unsafePerformIO` and also change `binDist` to make this so if it is preferred. Either way works.

Reply:
Yep please :) `unsafePerformIO` should not be used.